File: Diagrammatize.Design.Pattern/code/cxx/Command/A1/command/MacroCommand.cpp
#include "MacroCommand.hpp"

void MacroCommand::execute() {
    for (auto cmd : commands) {
        cmd->execute();
    }
}

void MacroCommand::append(std::shared_ptr<Command> cmd) {
    if (cmd.get() != this) {
        commands.push_back(cmd);
    }
}

void MacroCommand::undo() {
    if (!commands.empty()) {
        commands.pop_back();
    }
}

void MacroCommand::clear() {
    commands.clear();
}

File: Diagrammatize.Design.Patterns/code/python/Command/A1/command/MacroCommand.py
from .Command import Command

class MacroCommand(Command):
    def __init__(self):
        super().__init__()
        self.commands = list()

    def execute(self):
        for cmd in self.commands:
            cmd.execute()

    def append(self, cmd):
        if cmd != self:
            self.commands.append(cmd)

    def undo(self):
        if len(self.commands) > 0:
            self.commands.pop()

    def clear(self):
        self.commands = list()
